uacme: use acme-common and make it able to use acme.sh dns apis

orangepizza opened the pull request:
## 📦 Package Details

**Maintainer:** @lucize

**Description:**

- remake uacme hook based on acme-common's call structure, and allow acme virtual package can satisfied by uacme so it can be used with luci-app-acme. and I scrapped functions that used by acme.sh dnsapis and striped it's implementations into separate shell script, `dnsapihelper.sh`and linked into dns challenge hook, so uacme can use dns api from acme.sh without any modification from api side: they want curl or wget, and as uacme already depends on libcurl so I set curl as dependency so I changed dependency to acme-common +acme-acmesh-dnsapi if uacme is selected. while I copied most of acme.sh functions relevant form acme.sh, some if it's function is depend on OpenSSL. while re implemented some in ucode, some crypto functions like sign() and isecc etc will only work when you have OpenSSL, but as only about 10% of 304 scripts calls that I hope one can user themselves can install openssl if needed by their API.
-    while this does call ualpn server if config file is set to, it's user's duty to set ualpn server as needed as same as old version does. Not sure if it's safe to do this automatically because it'd need shutdown uhttpd to allow ualpn to listen port 443. original, same reason it doesn't support standalone validation, because uacme doesn't come with to do that.

---

## 🧪 Run Testing Details

- **OpenWrt Version:** snapshot
- **OpenWrt Target/Subtarget:** x86_64, but only things it touched is shell scripts so shouldn't matter
- **OpenWrt Device:** a kvm virtual machine

tested it can issue and renew certificate with uacme as backed, with config set from uacme without Openssl installed with duckdns /cloudflare/ webroot. (acme-common handles opening port when needed if I read code currectly) 

due to upstream bug in uacme, this well log a line of error in for not parsing ari info, but it's already fixed in their master branch and just waiting a new release for them. 
and uacme never rotate key by itself, until not-released yet version in master branch.

---

## ✅ Formalities

- [x] I have reviewed the [CONTRIBUTING.md](https://github.com/openwrt/packages/blob/master/CONTRIBUTING.md) file for detailed contributing guidelines.

orangepizza commented:
It does keep custom ca url (I tested on pebble entire time. See line 45 onword from hook.sh

 Tls-alpn-01 was never fully done (ulpn.sh was called bit never prepared to run its deamon)

mundschenk-at commented:
> It does keep custom ca url (I tested on pebble entire time. See line 45 onword from hook.sh

So what's the config key for that, `acme_server` (I've got no knowledge of `acme-acmesh` and the wiki at was silent on that at least in early 2024)? `uacme` uses `acme_url` (and `acme_url_staging`) since https://github.com/openwrt/packages/pull/23097. I'm not opposed to unifying that, but it needs documentation and/or a migration script.

> Tls-alpn-01 was never fully done (ulpn.sh was called bit never prepared to run its deamon)

It was implemented for `uacme` in https://github.com/openwrt/packages/pull/23097 (with manual `ualpn` setup as described in the PR).

orangepizza commented:
I assumed acmesh and uacme hook wouldn't diverged so just took scripts for acme-common from acme-acmesh and fit uacme command line options as that was what need for luci app acme

mundschenk-at commented:
Yeah, that's why I'm speaking up (it was a happy accident that I saw your PR, as this was my first and so far only code contribution to a package and I don't often visit this repo). Let's see how @lucize wants to resolve the URL issue.

stangri commented on net/acme-acmesh/Makefile:
```diff
@@ -51,13 +51,14 @@ endef
 
 define Package/acme-acmesh-dnsapi/description
  This package provides DNS API integration for ACME (Letsencrypt) client.
+ scripts are from acme.sh
```
The second line in the description does not conform to English grammar and isn't descriptive enough IMHO. Do you need any help redoing the description for this?

orangepizza replied:
I wanted to note that those functions are from acme.sh, as this PR make it can linked by other acme client too. But now not sure if it needed as it's name is acme-acmesh-dnsapi already:
Maby it'd better another package that links to this because uacmes dnsapi adapter got quite big (~30kb)

orangepizza committed:
acme: make virtual package satisfiable by uacme

currently acme metapackage only able to satisfied with acme-acmesh,
but make is satisfieable by uacme if it's already installed.
still defaults to acme.sh

Signed-off-by: Seo Suchan <tjtncks@gmail.com>

orangepizza committed:
acme-common: migrate uacme config paramaters

migrate old uacme specific config parameters into acme-common format.

Signed-off-by: Seo Suchan <tjtncks@gmail.com>

orangepizza committed:
uacme: use acme-common

remake uacme hook scripts to base on acme-common,
and implements helper to able to use acme.sh DNS APIs

Signed-off-by: Seo Suchan <tjtncks@gmail.com>

orangepizza committed:
uacme: update to 1.8.1

update uacme version to 1.8.1 , release upstream 2026/4/5

Signed-off-by: Seo Suchan <tjtncks@gmail.com>

tohojo reviewed (approved):
I have only looked at the shared bits, not the `uacme` changes themselves. These shared bits LGTM :)

orangepizza commented:
@commodo can this be merged or you would prefer minimal PR that just update uacme version?

commodo commented:
This looks like it went on long enough.
Merged.

Ping me for any breakage.

If possible, in the future consider adding some `test.sh` scripts.
In this era of AI assisted things, adding `test.sh` can become a higher priority (and not just rely on local testing).
I'm discovering quite a few broken dependencies in the `lang/python` space because of no `test.sh` scripts.

orangepizza commented:
Thanks for merging this. I didn't know about test.sh, will add if something else to fix.

mundschenk-at commented:
@orangepizza Can this and #29176 be cherry-picked for 25.12 to fix the mbedtls issue or does it rely on other infrastructure changes?

orangepizza commented:
@mundschenk-at I personally want to backport this to 25.12 base, but I think refactor itself is too big for backport. (as this touches acme-common too) 
think uacme version update itself can be considered bugfix, but that would be too different package at same version saparated by repo.
@commodo @tohojo  is it safe to backport this as is, tiny backport just update uacme binary version ( I think we now can because snapshot version is r2 now due to #29176 ) , or do not backport at all?

commodo commented:
> @mundschenk-at I personally want to backport this to 25.12 base, but I think refactor itself is too big for backport. (as this touches acme-common too) think uacme version update itself can be considered bugfix, but that would be too different package at same version saparated by repo. @commodo @tohojo is it safe to backport this as is, tiny backport just update uacme binary version ( I think we now can because snapshot version is r2 now due to #29176 ) , or do not backport at all?

If doing a version bump in a release branch I would also backport stuff.
Most of the times this shoukd be required.
